### ISSUE TITLE
Does not handle metadata that is a tagged literal

### ISSUE THREAD

**Issue by sogaiu** [sogaiu/tree-sitter-clojure#46]
As discovered [here](https://github.com/sogaiu/tree-sitter-clojure/issues/35#issuecomment-1445367208), it appears our grammar (421546c2547c74d1d9a0d8c296c412071d37e7ca) doesn't handle a construct which seems to be used by ClojureDart folks.

I've only managed to [find a bit under 150 `.cljd` files](https://github.com/sogaiu/tree-sitter-clojure/issues/35#issuecomment-1445006294) so far, but I found 8 files where a certain construct isn't handled (and the construct appears more than once in multiple files).

To repeat a bit here (new info follows the quote)...

__Begin Quote__

I found [something](https://github.com/kennytilton/flutter-mx/blob/ba934744a8a4f5589e5c329db61ca99e2aeb5464/src/tiltontec/flutter_mx/factory.cljd#L118-L133) our grammar doesn't seem to handle:
```
^#/(w/State ~fx-class #_m/SnackBar)
(...)
```

Assuming for the moment that that is valid, I presume the `^` indicates metadata.  [IIUC](https://github.com/sogaiu/tree-sitter-clojure/issues/35#issuecomment-1440122933), `#/(w/State ~fx-class #_m/SnackBar)` is a tagged literal.  So there is some metadata which is on `(...)` (the `...` is meant to stand in for more code here).

I guess that's not a case I anticipated: https://github.com/sogaiu/tree-sitter-clojure/blob/421546c2547c74d1d9a0d8c296c412071d37e7ca/grammar.js#L335-L342

If we're lucky a fix might be a matter of attending to that last `choice` there somehow.

Adding something in the last `choice` there (and I guess may be in `old_meta_lit`) is one option.

This brings up the topic of whether that `choice` should be there restricting things...possibly it could be replaced with `$._form`.

__End Quote__

I've tried both of the approaches mentioned above and they seem to work (though it may be that one is a bit slower than the other -- not sure yet).  I think they do ok on our clojars tests as well but I'll check that again after some sleep perhaps.

Below are sort of the raw error results that turned up:

* [flutter-mx.kennytilton/src/tiltontec/flutter_mx/factory.cljd](https://github.com/kennytilton/flutter-mx/blob/ba934744a8a4f5589e5c329db61ca99e2aeb5464/src/tiltontec/flutter_mx/factory.cljd#L118)	3 ms	(ERROR [117, 21] - [117, 22])
* [flutter-mx.kennytilton/src/tiltontec/example/x020_reactive_stream.cljd](https://github.com/kennytilton/flutter-mx/blob/ba934744a8a4f5589e5c329db61ca99e2aeb5464/src/tiltontec/example/x020_reactive_stream.cljd#L86)	1 ms	(ERROR [85, 56] - [85, 57])
* [flutter-mx.kennytilton/src/tiltontec/example/x05_fab.cljd](https://github.com/kennytilton/flutter-mx/blob/ba934744a8a4f5589e5c329db61ca99e2aeb5464/src/tiltontec/example/x05_fab.cljd#L176)	1 ms	(ERROR [175, 10] - [175, 11])
* [flutter-mx.kennytilton/src/tiltontec/example/x03_physics_sim.cljd](https://github.com/kennytilton/flutter-mx/blob/ba934744a8a4f5589e5c329db61ca99e2aeb5464/src/tiltontec/example/x03_physics_sim.cljd#L18)	0 ms	(ERROR [17, 18] - [17, 19])
* [ClojureDart.Tensegritics/samples/fab/src/sample/fab.cljd](https://github.com/Tensegritics/ClojureDart/blob/60bd59262d24218ac5e24c2d9d4e49a4f1b969be/samples/fab/src/sample/fab.cljd#L46)	1 ms	(ERROR [45, 44] - [45, 45])
* [ClojureDart.Tensegritics/clj/src/cljd/core.cljd](https://github.com/Tensegritics/ClojureDart/blob/60bd59262d24218ac5e24c2d9d4e49a4f1b969be/clj/src/cljd/core.cljd#L2650)	46 ms	(ERROR [2649, 6] - [2649, 7])
* [ClojureDart.Tensegritics/clj/src/cljd/reader.cljd](https://github.com/Tensegritics/ClojureDart/blob/60bd59262d24218ac5e24c2d9d4e49a4f1b969be/clj/src/cljd/reader.cljd#L5)	5 ms	(ERROR [4, 22] - [4, 23])
* [ClojureDart.Tensegritics/clj/test/cljd/test_clojure/core_test_cljd.cljd](https://github.com/Tensegritics/ClojureDart/blob/60bd59262d24218ac5e24c2d9d4e49a4f1b969be/clj/test/cljd/test_clojure/core_test_cljd.cljd#L356)	5 ms	(ERROR [355, 24] - [355, 25])

**Comment by sogaiu** [comment]
Thanks for your comments.

>  Would ^ followed by #/( ... ) be considered a single form?

Yes.  Though there are a number of things going on here which I think might lead to confusion.

`^#/(<some-stuff>)` is metadata and it is on something -- which in the original example above I wrote as `(...)`.

My impression is that tagged literals are often written with a space after the symbol portion, so another way to express things is:

```
^#/ (<some-stuff>) (...)
```

Here there is a space after the `/` (which is a symbol).  

This way of articulating things is closer to the more commonly encountered (at least for me) uuid case:

```
#uuid "00000000-0000-0000-0000-000000000000"
```

FWIW, the first tweak I tried with the grammar was:
```
    meta_lit: $ =>
    seq(field('marker', "^"),
        repeat($._gap),
        field('value', choice($.read_cond_lit,
                              $.tagged_or_ctor_lit,
                              $.map_lit,
                              $.str_lit,
                              $.kwd_lit,
                              $.sym_lit))),
```

With that change, using ahlinc's alpha, I get:
```
$ echo "^#/(data) (thing)" | tree-sitter.ahlinc parse -
(source [0, 0] - [1, 0]
  (list_lit [0, 0] - [0, 17]
    meta: (meta_lit [0, 0] - [0, 9]
      value: (tagged_or_ctor_lit [0, 1] - [0, 9]
        tag: (sym_lit [0, 2] - [0, 3]
          name: (sym_name [0, 2] - [0, 3]))
        value: (list_lit [0, 3] - [0, 9]
          value: (sym_lit [0, 4] - [0, 8]
            name: (sym_name [0, 4] - [0, 8])))))
    value: (sym_lit [0, 11] - [0, 16]
      name: (sym_name [0, 11] - [0, 16]))))
```

**Comment by sogaiu** [comment]
So it should be possible to test things out by using the `dev` branch at [this commit](https://github.com/sogaiu/tree-sitter-clojure/commit/0f4e0649445b36b2f38cbfc63cd03be28e926395).

If you have an appropriate version of babashka, it should be possible to use babashka tasks that will fetch ClojureDart code as well as run the `parse` subcommand on the fetched results.

To see a list of tasks try `bb tasks` -- that should show the things listed here: https://github.com/sogaiu/tree-sitter-clojure/blob/0f4e0649445b36b2f38cbfc63cd03be28e926395/bb.edn#L5-L6

Execution is just `bb fetch-code` and `bb parse-tree`.

The corresponding tasks are defined in files that live [here](https://github.com/sogaiu/tree-sitter-clojure/tree/0f4e0649445b36b2f38cbfc63cd03be28e926395/script).

**Comment by sogaiu** [comment]
As we are [splitting off some of the dev-related bits into another repository](https://github.com/sogaiu/tree-sitter-clojure/issues/49#issuecomment-1464799429) and some of those bits were on the dev branch, I've made a separate pre-0.0.12 branch to preserve the bits that are remaining here.

The fix can now be seen at 7df2726280aef74ed963d980c1fa9c9f554eac97.

**Comment by sogaiu** [comment]
As mentioned above, addressed in 7df2726.
